### PR TITLE
Mini Project 0

### PR DESCRIPTION
**Pull request by Rshah2067**

### CODE REVIEW

File: mini-projects/00-jq-rs/Cargo.toml [hunk 1]
```diff
@@ -4,3 +4,8 @@ version = "0.1.0"
 edition = "2021"
 
 [dependencies]
+clap = {version = "4.5.20", features = ["derive"]}
+regex = "1.11.0"
+serde_json ={version = "1.0.128", features = ["preserve_order"]}
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/Cargo.toml:9]: nice job finding this feature

File: mini-projects/00-jq-rs/src/main.rs [hunk 2]
```diff
@@ -1,3 +1,61 @@
...
+    path: String,
+}
+#[derive(Error, Debug)]
+pub enum ParseError {
+    #[error("File error: {0}")]
+    FileError(#[from] io::Error), // Automatically convert from std::io::Error
+
+    #[error("JSON parsing error: {0}")]
+    JsonError(#[from] serde_json::Error), // Automatically convert from serde_json::Error
+}
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/main.rs:31]: nice use of this error

File: mini-projects/00-jq-rs/src/main.rs [hunk 3]
```diff
@@ -1,3 +1,61 @@
...
 fn main() {
-    println!("Hello World!");
+    let args = Args::parse();
+    let jq_colors: std::result::Result<String, env::VarError> = env::var("JQ_COLORS");
+    //check if we found an env variable
+    let jq_env = match jq_colors {
+        Ok(env) => env,
+        //default value is used if one is not found
+        Err(_) => String::from("0;90:0;37:0;37:0;37:0;32:1;37:1;37:1;34"),
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/main.rs:39]: nit: this string is a little gross, potentially would have been nice to assign each individual parameter to its own constant and then just format it into this string - that way, a dev can quickly scan the variables names to know which one is relevant to them, rather than needing to pull up the documentation

File: mini-projects/00-jq-rs/src/main.rs [hunk 4]
```diff
@@ -1,3 +1,61 @@
...
+        args.needle,
+        json,
+        args.coloroutput,
+        args.sortkeys,
+        args.indent,
+        args.compactoutput,
+        &jq_env,
+    );
+}
+//TODO add error handling
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/main.rs:56]: seems like this todo was completed?

File: mini-projects/00-jq-rs/src/processor.rs [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+    for command in list {
+        //if the input for the next stage is anything other than a value we should panic and tell the user that they gave an invalid filter
+        match input {
+            Result::Value(ref val) => input = execute(command.0, command.1, val.clone()),
+            _ => panic!("Invalid input filter!"),
+        };
+    }
+    input
+}
+pub fn call_printer(
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/processor.rs:99]: super nit: this function seems slightly misnamed - I would think `call_printer` would only print, not also run the filters

File: mini-projects/00-jq-rs/src/processor/filter.rs [hunk 6]
```diff
@@ -0,0 +1,163 @@
+use serde_json::Value;
+use std::ops::Range;
+use std::num::ParseIntError;
+use thiserror::Error;
+//The pipe and iterator filter modify the way filters are executed and aren't treated as filters
+#[derive(PartialEq, Debug)]
+pub enum FilterType {
+    Identity,
+    ObjectIdentity,
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/processor/filter.rs:9]: identity seems confusing as opposed to identifier

File: mini-projects/00-jq-rs/src/processor/operations.rs [hunk 7]
```diff
@@ -0,0 +1,214 @@
+use serde_json::Value;
+//Enum representing the different outputs we could get from an operation
+#[derive(PartialEq, Debug)]
+pub enum Result {
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/processor/operations.rs:4]: seems confusing to name it a "Result"

File: mini-projects/00-jq-rs/src/processor.rs [hunk 8]
```diff
@@ -0,0 +1,253 @@
+use filter::*;
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/processor.rs:1]: not the biggest fan of * imports, makes it less explicit where things come from

File: mini-projects/00-jq-rs/src/processor.rs [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+            ".[]" => output.push((Commands::ArrayIterator, String::new())),
+            //now we want to search for commands like delete[], delete<>,array index,."key" and array slice
+            //most convient way to do this I found was to use REGEX
+            _ => {
+                //no need for errorhandling as the regex is not userfacing
+                let delete_key_regex = Regex::new(r"del\(\.(.*?)\)").unwrap();
+                let delete_index_regex = Regex::new(r"del\[\.(.*?)\]").unwrap();
+                let array_slice_regex = Regex::new(r"\.\[([^\]]*:[^\]]*)\]").unwrap();
+                let array_index_regex = Regex::new(r"\.\[(.*?)\]").unwrap();
+                let object_regex = Regex::new(r"\.(\w+)").unwrap();
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/processor.rs:42]: this is a great use of regex

File: mini-projects/00-jq-rs/src/processor.rs [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+                                                Commands::Filter(FilterType::ArrayIndex),
+                                                String::from(&needle[1]),
+                                            )),
+                                            None => {
+                                                //finally we can check if it is an object index, if it isn't this than the input was invalid so the code will panic
+                                                let needle = object_regex.captures(string).unwrap();
+                                                output.push((
+                                                    Commands::Filter(FilterType::ObjectIdentity),
+                                                    String::from(&needle[1]),
+                                                ))
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/processor.rs:72]: lol this verbosely nesting is kinda gross, in this case, I probably would have opted for if and elifs

File: mini-projects/00-jq-rs/src/processor/printer.rs [hunk 11]
```diff
@@ -0,0 +1,424 @@
...
+    };
+}
+fn print_value(
+    input: Value,
+    color: bool,
+    sort: bool,
+    indent: u32,
+    compact: bool,
+    jq_vec: Vec<(&str, &str)>,
+) -> String {
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/processor/printer.rs:40]: I think having a print trait that multiple different print structs imply could have cleaned up some of this logic - without it, seems like there's a lot of branching within each subsection of this function

File: mini-projects/00-jq-rs/src/processor/operations.rs [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+impl Delete {
+    pub fn new(input: Value, needle: String) -> Self {
+        let output = Delete {
+            input: input,
+            needle: needle,
+        };
+        output
+    }
+}
+impl Operate for Delete {
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/processor/operations.rs:61]: I like this - Operate seems like a somewhat poor naming choice - I know what your going for, but operate is a pretty ambiguous term

**Comment by ayushchakra** [comment]
# Rubric For Mini Project 0: JQ

## Submission

### Canvas

1: The link submitted on Canvas is for the pull request containing the
assignment submission.

0: The link submitted on Canvas is not for the pull request containing the
assignment submission.

### Git Branch

1: The branch includes only changes for this assignment.

0: The branch includes commits or changes outside this assignment.

### Git Commit Message

1: All commit messages for the branch are reasonably clear.

0: One or more commit messages for the branch are reasonably clear.

### GitHub Pull Request

1: The pull request is made to the `main` branch of the student's fork of the
course repository.

0: The pull request is made to a branch other than `main` or to a repo that is
not the student's fork of the course repository.

### Points

**Submission Score: 4/4**

## Features
### Command Line Interface
For each of the required CLI arguments `filter-string`, `path-to-json`, `color-output`, `monochrome-output`, `sort-keys`, `indent`, and `compact-output`

0.5: The argument can be correctly parsed and consumed by the rest of the package.

0: The argument is not correctly parsed and propagated to the rest of the package.

Additionally, the command line parsing correctly raises errors to the user when conflicting arguments are inputted.

1.5: Errors are properly determined and bubbled up to the user.

0: Errors are not propagated back to the user.

#### Comments
-1.5 for not raising an error for conflicting arguments

### Supported Filters
For each of the required filters `identity`, `object-identifier index`, `array index`, `array slice`, `pipe`, and `array iterator`

2: The filter is correctly implemented and properly parses all possible JSON inputs.

1: Minor errors exist in the implement of the filter.

0: Major errors exist in the implementation, or the filter was omitted from the package.

Additionally, comprehensive unit tests are required for each feature:

2: Full unit test coverage is provided (with passing implementations) to exercise a wide varity of use-cases and edge-cases of the filter, ensuring that the output is correct and that errors are properly outputted.

1: Unit tests are present, but lack a significant breadth of cases to fully exercise the filter.

0: Tests are missing, or lack significant assessment of the filter's performance.

#### Comments
Identity: Correct implementation and unit test
Object Identifier: Correct implementation and unit test
Array Index: Correct implementation and unit test
Array Slice: Correct implementation and unit test
Pipe: Correct implementation und unit test
(-2) Array iterator: Not implemented

(-2) Not going to doc points everywhere, but as a whole, your tests don't cover the negative case for where you expect an error due to an invalid input. Also, it would have been good to test that after the filter was correctly identified, it would also apply the correct operation or if there was a bug in the pipeline

### Built-In Functions
For each of the required built-in functions `add`, `length`, `del`

2: The built-in function is correctly implemented and properly processes all possible JSON inputs.

1: Minor errors exist in the implementation of the operator.

0: Major errors exist in the implementation, or the function was omitted from the package.

Additionally, comprehensive unit tests are required for each function:

2: Full unit test coverage is provided (with passing implementations) to exercise a wide varity of use-cases and edge-cases of the function, ensuring that the output is correct and that errors are properly outputted.

1: Unit tests are present, but lack a significant breadth of cases to fully exercise the function.

0: Tests are missing, or lack significant assessment of the function's performance.

#### Comments
Add: Correct Implementation and Unit tests
Length: Correct implementation and unit tests
Del: Correct implementation and unit tests

(-2) Similar to above, your test cases don't assess the negative cases.

### Formatted Printing
For each of the required formats, `colored printing`, `monochrome printing`, `sorted keys`, `indents`, and `compact output`

2: The formatter is correctly implemented and properly displays any JSON object..

1: Minor errors exist in the implementation of the formatter.

0: Major errors exist in the implementation, or the formatter was omitted from the package.

Additionally, unit tests are required for each function:

2: Full unit test coverage is provided (with passing implementations) to ensure that the formatter can correctly handle all expected inputs.

1: Unit tests are present, but lack significant breadth of cases to fully exercise the formatter.

0: Tests are missing, or lack significant assessment of the formatter.

#### Comments
Colored Printing: Correctly implemented and unit tests
(-2) Monochrome Printing: Not implemented
(-2) Sorted Keys: Correctly implemented, but no unit tests
(-3) Indents: There is a minor error in the indent. The code does not properly account for nested complex datatypes, which would require multiple layers of iterating and no unit tests
(-3) Compact Output: Similarly, there is a minor error that still places new brackets on a newline

### Points
**Features Score: 44/61**

## Code Quality

### Design

#### File/Module organization

2: The program is appropriately split into reasonable files/modules

1: The program is somewhat organized, but there's room for improvement (splits aren't idiomatic, or some modules are still too large, etc.)

0: The program is not split at all

#### Scalability

5: The code is architected with scalability in mind. If more features needed to be added to the filtering/printing/CLI, they could easily be integrated.

3: The code was architected with some scalaibility in mind. However, significant refactoring would be needed in order to expand the feature set of this tool. (Possible symptoms include not modularizing enough/not effectively using traits + generics).

0: The code is not scalable. The entire package would need to be refactored in order to expand its features.

#### General Cleanliness

4: The code is effectively documented, especially for sources of user-facing errors. Additionally, function signatures are clear and the code is overall very readable.

2: Attempts at verbose documentation were made, but certain sources of error are unclear OR poor function signatures are used and readers would have slight difficulties while scanning the code.

0: Documentation is severely lacking in several palces and the code is not written in a readable way.

### Lint/Format

#### RustFmt

2: RustFmt produces no warnings when run on the changed files

1: RustFmt produces a few minor warnings when run on the changed files

0: RustFmt produces many warnings when run on the changed files

#### Clippy

2: Clippy produces no warnings when run on the changed files

1: Clippy produces a few minor warnings when run on the changed files

0: Clippy produces many warnings when run on the changed files

### Points
**Code Quality Score: 10/15**

#### Comments
Docing one point for scalability. You definitely did some great stuff in your filtering to make it nice a generic, but I think some work can be done to improve the printing scalability.
Also docing one point for general cleanliness, I thought some of your errors could have been more verbose, and also could have added slightly more documentation for longer logical blocks.
-1 for rustfmt
-2 for all the clips warnings

## Summary
Add up the point totals here

**Total Score: 58/80**
